File: ols-shared/src/main/java/uk/ac/ebi/ols/shared/DefinedFields.java
package uk.ac.ebi.ols.shared;

public enum DefinedFields {
    APPEARS_IN("appearsIn", "", "The list of ontologies in which the current entity is used.", "array"),
    BASE_URI("baseUri", "baseUris",
            "The beginning of URIs that OLS assumes to belong to an ontology.", "array"),
    DEFINED_BY("definedBy", "", "A list of ontologies which defines this term", "array"),
    DEFINITION("definition", "description","The definition of this entity.", "string"),
    DIRECT_ANCESTOR("directAncestor", "",
            "A list of all direct parents by subclass- and hierarchical relations for this entity.", "array"),
    DIRECT_PARENT("directParent", "","A list of the direct parents of this entity.", "array"),
    EXPORTS_TO("exportsTo", "", "Lists the prefixes of ontologies that imports this ontology.", "array"),
    HAS_DIRECT_CHILDREN("hasDirectChildren", "has_children",
            "Whether this class has direct children or not.", "boolean"),
    HAS_DIRECT_PARENTS("hasDirectParents", "",
            "Indicates whether this class has direct parents or not.", "boolean"),
    HAS_HIERARCHICAL_CHILDREN("hasHierarchicalChildren", "",
            "Whether this class has hierarchical children or not.", "boolean"),
    HAS_HIERARCHICAL_PARENTS("hasHierarchicalParents", "",
            "Whether this class has hierarchical parents or not.", "boolean"),
    HAS_INDIVIDUALS("hasIndividuals", "", "Whether individuals exists or not.", "boolean"),
    HAS_LOCAL_DEFINITION("hasLocalDefinition", "",
            "True if term is definined within this ontology.", "boolean"),
    HIERARCHICAL_ANCESTOR("hierarchicalAncestor","","The list of ancestors of this entity via " +
            "subclass relationships and hierarchical properties such as part_of(BFO:0000050) relations.", "array"),
    HIERARCHICAL_PARENT("hierarchicalParent","" ,"The list of parents of this entity via " +
            "subclass relationships and hierarchical properties such as part_of(BFO:0000050) relations.", "array"),
    IMPORTED("imported", "",
            "Whether this entity is imported or not.", "boolean"),
    IMPORTS_FROM("importsFrom", "",
            "Lists the prefixes of the ontologies the current ontology imports from.", "array"),
    IS_DEFINING_ONTOLOGY("isDefiningOntology",
            "is_defining_ontology","Whether this entity is defined in this ontology or not.", "boolean"),
    IS_OBSOLETE("isObsolete", "isObsolete",
                        "Set to true if this entity is obsolete, otherwise is set to false.", "boolean"),
    IS_PREFERRED_ROOT("isPreferredRoot", "is_preferred_root",
            "Set to true if this entity is a preferred root or not.", "boolean"),
    LABEL("label", "", "The name or names of this entity.", "string"),
    LANGUAGE("language", "", "The language or languages this ontology is available in.", "array"),
    MAILING_LIST("mailingList", "", "The mailing list for the ontology.", "string"),
    NUM_DESCENDANTS ("numDescendants", "", "Number of descendants of this entity.", "integer"),
    NUM_HIERARCHICAL_DESCENDANTS ("numHierarchicalDescendants", "",
            "Number of hierarchical descendants of this entity.", "integer"),
    ONTOLOGY_PURL("ontologyPurl", "", "The URL of the ontology to download the ontology file.", "string"),
    PREFERRED_ROOT("preferredRoot", "",
            "A list of entities that serve as roots for this ontology.", "array"),
    RELATED_FROM("relatedFrom","", "The list of classes in which this class is used as part of its definition. " +
            "I.e. 'liver disease' (EFO:0001421) has a relatedFrom relation with 'serum albumin measurement' (EFO:0004535) because " +
            "'serum albumin measurement' is a subclass of 'is about some liver disease'.", "array"),
    RELATED_TO("relatedTo","", "The list of classes that are used in axioms in the filler position" +
            " in defining this class. It only considers classes in the filler position, not classes expressions in the filler position." +
            "I.e. 'liver disease' (EFO:0001421) has a relatedTo relation with 'liver' (UBERON:0002107) because " +
            "'liver disease' is a subclass of 'has_disease_location some liver'.", "array"),
    SYNONYM("synonym", "", "The list of names that are synonyms of this entity.", "array");

    private final String text;
    private final String ols3Text;
    private final String description;
    private final String type;

    DefinedFields(String text, String ols3Text, String description, String type) {
        this.text = text;
        this.ols3Text = ols3Text;
        this.description = description;
        this.type = type;
    }

    public String getText() {
        return text;
    }

    public String getOls3Text() {
        return ols3Text;
    }

    public String getDescription() { return description; }

    public String getType() { return type; }
}

File: dataload/ols-shared-rs/src/defined_fields.rs
/// Defined fields matching the Java DefinedFields enum
/// These are the standard field names used across OLS4 JSON processing
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[allow(dead_code)]
pub enum DefinedFields {
    AppearsIn,
    BaseUri,
    DefinedBy,
    Definition,
    DirectAncestor,
    DirectParent,
    ExportsTo,
    HasDirectChildren,
    HasDirectParents,
    HasHierarchicalChildren,
    HasHierarchicalParents,
    HasIndividuals,
    HasLocalDefinition,
    HierarchicalAncestor,
    HierarchicalParent,
    Imported,
    ImportsFrom,
    IsDefiningOntology,
    IsObsolete,
    IsPreferredRoot,
    Label,
    Language,
    MailingList,
    NumDescendants,
    NumHierarchicalDescendants,
    OntologyPurl,
    PreferredRoot,
    RelatedFrom,
    RelatedTo,
    LinksTo,
    Synonym,
}

impl DefinedFields {
    /// Get the JSON field name (primary name)
    pub fn text(&self) -> &'static str {
        match self {
            DefinedFields::AppearsIn => "appearsIn",
            DefinedFields::BaseUri => "baseUri",
            DefinedFields::DefinedBy => "definedBy",
            DefinedFields::Definition => "definition",
            DefinedFields::DirectAncestor => "directAncestor",
            DefinedFields::DirectParent => "directParent",
            DefinedFields::ExportsTo => "exportsTo",
            DefinedFields::HasDirectChildren => "hasDirectChildren",
            DefinedFields::HasDirectParents => "hasDirectParents",
            DefinedFields::HasHierarchicalChildren => "hasHierarchicalChildren",
            DefinedFields::HasHierarchicalParents => "hasHierarchicalParents",
            DefinedFields::HasIndividuals => "hasIndividuals",
            DefinedFields::HasLocalDefinition => "hasLocalDefinition",
            DefinedFields::HierarchicalAncestor => "hierarchicalAncestor",
            DefinedFields::HierarchicalParent => "hierarchicalParent",
            DefinedFields::Imported => "imported",
            DefinedFields::ImportsFrom => "importsFrom",
            DefinedFields::IsDefiningOntology => "isDefiningOntology",
            DefinedFields::IsObsolete => "isObsolete",
            DefinedFields::IsPreferredRoot => "isPreferredRoot",
            DefinedFields::Label => "label",
            DefinedFields::Language => "language",
            DefinedFields::MailingList => "mailingList",
            DefinedFields::NumDescendants => "numDescendants",
            DefinedFields::NumHierarchicalDescendants => "numHierarchicalDescendants",
            DefinedFields::OntologyPurl => "ontologyPurl",
            DefinedFields::PreferredRoot => "preferredRoot",
            DefinedFields::RelatedFrom => "relatedFrom",
            DefinedFields::RelatedTo => "relatedTo",
            DefinedFields::LinksTo => "linksTo",
            DefinedFields::Synonym => "synonym",
        }
    }

    /// Get the alternative name (if any) - used for legacy compatibility
    pub fn alt(&self) -> &'static str {
        match self {
            DefinedFields::BaseUri => "baseUris",
            DefinedFields::Definition => "description",
            DefinedFields::HasDirectChildren => "has_children",
            _ => "",
        }
    }
}

/// Constants for commonly used field names
pub mod fields {
    pub const BASE_URI: &str = "baseUri";
    pub const DEFINED_BY: &str = "definedBy";
    pub const DEFINITION: &str = "definition";
    pub const DIRECT_ANCESTOR: &str = "directAncestor";
    pub const DIRECT_PARENT: &str = "directParent";
    pub const HIERARCHICAL_ANCESTOR: &str = "hierarchicalAncestor";
    pub const HIERARCHICAL_PARENT: &str = "hierarchicalParent";
    pub const IMPORTED: &str = "imported";
    pub const IS_DEFINING_ONTOLOGY: &str = "isDefiningOntology";
    pub const IS_OBSOLETE: &str = "isObsolete";
    pub const IS_PREFERRED_ROOT: &str = "isPreferredRoot";
    pub const LABEL: &str = "label";
    pub const LANGUAGE: &str = "language";
    pub const NUM_DESCENDANTS: &str = "numDescendants";
    pub const NUM_HIERARCHICAL_DESCENDANTS: &str = "numHierarchicalDescendants";
    pub const ONTOLOGY_PURL: &str = "ontologyPurl";
    pub const PREFERRED_PREFIX: &str = "preferredPrefix";
    pub const PREFERRED_ROOT: &str = "preferredRoot";
    pub const SYNONYM: &str = "synonym";
}
